llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -2626,6 +2671,151 @@ Register SPIRVInstructionSelector::buildPointerToResource(
                                                  MIRBuilder);
 }
 
+bool SPIRVInstructionSelector::selectFirstBitHigh16(Register ResVReg,
+                                                    const SPIRVType *ResType,
+                                                    MachineInstr &I,
+                                                    bool IsSigned) const {
+  unsigned Opcode = IsSigned ? SPIRV::OpSConvert : SPIRV::OpUConvert;
+  // zero or sign extend
+  Register ExtReg = MRI->createVirtualRegister(GR.getRegClass(ResType));
+  bool Result =
+      selectUnOpWithSrc(ExtReg, ResType, I, I.getOperand(2).getReg(), Opcode);
+  return Result & selectFirstBitHigh32(ResVReg, ResType, I, ExtReg, IsSigned);
+}
+
+bool SPIRVInstructionSelector::selectFirstBitHigh32(Register ResVReg,
+                                                    const SPIRVType *ResType,
+                                                    MachineInstr &I,
+                                                    Register SrcReg,
+                                                    bool IsSigned) const {
+  unsigned Opcode = IsSigned ? GL::FindSMsb : GL::FindUMsb;
+  return BuildMI(*I.getParent(), I, I.getDebugLoc(), TII.get(SPIRV::OpExtInst))
```

Comment:
Seems like this could just be `selectExtInst`  instead of `BuildMI`.

Reply:
If I used 'selectExtInst' it would only use the 2nd operand as the SrcReg, but the way this function is used by the 16bit and 64bit functions, it needs more freedom in what the SrcReg is.

Reply:
Fair enough. Alternatively, that sounds like the 2nd operand could be updated to the srcReg before calling `selectFirstBitHigh32` and then we don't pass a  srcRegister to `selectFirstBitHigh32`.

Reply:
It didn't occur to me to update the operand, so that is possible. I guess it depends on what will be more understandable.

Reply:
Unfortunately, there are no MIR tests in this backend for me to know if this is a problem or not, but I was curious if we don't update a stale the operand is that's essentially preserving it?

Reply:
Nvmd upon looking only llvm IR has a `setOperand` I don't see one in MIR so this is fine.